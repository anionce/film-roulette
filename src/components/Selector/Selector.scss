@import '../../assets/styles/variables';

.root-element {
	font-size: 20;
	font-family: $main-font;
	width: $sm;
	height: 60px;
	padding: 0;
	border-radius: 4px;
	border: 1px solid $color-black;
	background-color: white;

	.MuiOutlinedInput-notchedOutline {
		border-style: none;
	}

	.MuiSelect-outlined.MuiSelect-select {
		min-height: auto;
		font-size: $font-size-md;
	}
}

@media only screen and (max-width: $lg) {
	.root-element {
		max-width: 300px;
		max-height: 40px;
	}
}

@media only screen and (max-width: $sm) {
	.root-element {
		max-width: 200px;
		max-height: 30px;

		.MuiSelect-outlined.MuiSelect-select {
			font-size: $font-size-sm;
		}
	}
}
