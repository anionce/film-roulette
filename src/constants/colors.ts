export const BACKGROUND_GREY = '#E5E5E5';
