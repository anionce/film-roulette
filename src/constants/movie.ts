export const MOVIE_POSTER_PATH = 'https://image.tmdb.org/t/p/original/';

export const MOVIE_IMDB_PATH = 'https://www.imdb.com/title/';
