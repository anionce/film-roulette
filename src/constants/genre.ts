import { SelectOptions } from './selector';

export const mapValueToGenre = (value: MovieGenre): number[] | null => {
	if (GENRE_NUM[value]) {
		return [GENRE_NUM[value] as number];
	}
	return null;
};

export enum MovieGenre {
	Action = 'acción',
	Adventure = 'aventuras',
	Animation = 'animación',
	Comedy = 'comedia',
	Crime = 'crimen',
	Documentary = 'documental',
	Drama = 'drama',
	Family = 'familiar',
	Fantasy = 'fantasía',
	History = 'histórica',
	Horror = 'terror',
	Music = 'música',
	Mystery = 'misterio',
	Romance = 'romance',
	ScienceFiction = 'ciencia ficción',
	Thriller = 'thriller',
	War = 'bélica',
	Western = 'western',
	Random = 'random',
}

export const GENRE_NUM: Record<MovieGenre, number | null> = {
	[MovieGenre.Action]: 28,
	[MovieGenre.Adventure]: 12,
	[MovieGenre.Animation]: 16,
	[MovieGenre.Comedy]: 35,
	[MovieGenre.Crime]: 80,
	[MovieGenre.Documentary]: 99,
	[MovieGenre.Drama]: 18,
	[MovieGenre.Family]: 10751,
	[MovieGenre.Fantasy]: 14,
	[MovieGenre.History]: 36,
	[MovieGenre.Horror]: 27,
	[MovieGenre.Music]: 10402,
	[MovieGenre.Mystery]: 9648,
	[MovieGenre.Romance]: 10749,
	[MovieGenre.ScienceFiction]: 878,
	[MovieGenre.Thriller]: 53,
	[MovieGenre.War]: 10752,
	[MovieGenre.Western]: 37,
	[MovieGenre.Random]: null,
} as const;

export const genreSelectorOptions: SelectOptions[] = Object.keys(GENRE_NUM).map(genre => ({
	value: genre,
	text: genre,
}));
