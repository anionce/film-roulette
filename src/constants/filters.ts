import { MovieGenre } from './genre';
import { MovieRuntime } from './runtime';
import { StreamingServices } from './streamingServices';

export enum FilterType {
	Duration = 'duration',
	Genre = 'genre',
	Streaming = 'streaming',
}

export const filterTypes = [FilterType.Genre, FilterType.Duration, FilterType.Streaming];

export type ModalOpen = {
	[FilterType.Genre]: boolean;
	[FilterType.Duration]: boolean;
	[FilterType.Streaming]: boolean;
};

export type FilterArguments = {
	genre: MovieGenre | null;
	duration: MovieRuntime | null;
	streaming: StreamingServices[] | null;
};
