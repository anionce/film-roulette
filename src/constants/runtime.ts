import { SelectOptions } from './selector';

export const mapValueToMovieRuntime = (value: MovieRuntime): number => {
	return RUNTIME_LENGTH[value];
};

export enum MovieRuntime {
	Short = 'short',
	Medium = 'medium',
	Long = 'long',
}

export const runtimeSelectorOptions: SelectOptions[] = [
	{ value: MovieRuntime.Short, text: '1.30h' },
	{ value: MovieRuntime.Medium, text: '2h' },
	{ value: MovieRuntime.Long, text: 'Plenty of time' },
];

export const RUNTIME_LENGTH: Record<MovieRuntime, number> = {
	[MovieRuntime.Short]: 95,
	[MovieRuntime.Medium]: 125,
	[MovieRuntime.Long]: 200,
} as const;
