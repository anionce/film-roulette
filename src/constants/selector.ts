import { FilterType } from './filters';
import { genreSelectorOptions } from './genre';
import { runtimeSelectorOptions } from './runtime';

export type SelectOptions = {
	value: string;
	text: string;
};

export const getOptionsForSelector = (value: FilterType) => {
	if (value === FilterType.Genre) {
		return genreSelectorOptions;
	} else {
		return runtimeSelectorOptions;
	}
};
