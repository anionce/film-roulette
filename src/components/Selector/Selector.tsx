import React from 'react';
import './Selector.scss';
import MenuItem from '@mui/material/MenuItem';
import Select, { SelectChangeEvent } from '@mui/material/Select';
import { getOptionsForSelector, SelectValue } from '../../constants/selector';
import { MovieRuntime } from '../../constants/runtime';
import { MovieGenre } from '../../constants/genre';

export type SelectorProps = {
	selectorValue: SelectValue;
	onChange: (event: SelectChangeEvent<MovieGenre | MovieRuntime>) => void;
	selectedOption: MovieGenre | MovieRuntime;
};

export const Selector = ({ selectorValue, onChange, selectedOption }: SelectorProps) => {
	return (
		<Select value={selectedOption} onChange={onChange} classes={{ root: 'root-element' }}>
			{getOptionsForSelector(selectorValue).map(({ text, value }) => (
				<MenuItem
					sx={{ minHeight: 35, fontFamily: 'Playfair Display', fontSize: 20 }}
					classes={{ root: 'menu-item' }}
					key={value}
					data-testid={value}
					value={value}>
					{text}
				</MenuItem>
			))}
		</Select>
	);
};
