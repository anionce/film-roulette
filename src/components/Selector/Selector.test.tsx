import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { customRender } from '../../assets/tests/testUtils';
import { Selector, SelectorProps } from './Selector';
import { SelectValue } from '../../constants/selector';
import { MovieGenre } from '../../constants/genre';
import { fireEvent, screen } from '@testing-library/react';
import { MovieRuntime, runtimeSelectorOptions } from '../../constants/runtime';

const defaultProps: SelectorProps = {
	selectorValue: SelectValue.Genre,
	onChange: jest.fn(),
	selectedOption: MovieGenre.Action,
};

test('renders correcly', () => {
	const { container } = customRender(<Selector {...defaultProps} />);

	expect(container).toMatchSnapshot();
});

test('when clicking on select with duration, all options are displayed', () => {
	customRender(
		<Selector {...defaultProps} selectorValue={SelectValue.Runtime} selectedOption={MovieRuntime.Short} />
	);

	fireEvent.mouseDown(screen.getByText(runtimeSelectorOptions[0].text));

	const durations = runtimeSelectorOptions
		.map(runtime => runtime.text)
		.filter(runtime => runtime !== runtimeSelectorOptions[0].text);

	durations.forEach(duration => {
		expect(screen.getByText(duration)).toBeInTheDocument();
	});
});

test('when clicking on select with genre, all options are displayed', () => {
	customRender(<Selector {...defaultProps} />);

	fireEvent.mouseDown(screen.getByText(defaultProps.selectedOption));

	const genres = Object.values(MovieGenre).filter(genre => genre !== defaultProps.selectedOption);

	genres.forEach(genre => {
		expect(screen.getByText(genre)).toBeInTheDocument();
	});
});

test('when selecting an option, onChange should be called and the option will be displayed', () => {
	customRender(<Selector {...defaultProps} />);

	fireEvent.mouseDown(screen.getByText(defaultProps.selectedOption));

	fireEvent.click(screen.getByText(MovieGenre.Comedy));

	expect(defaultProps.onChange).toHaveBeenCalled();

	expect(screen.getByText(MovieGenre.Comedy)).toBeInTheDocument();
});
