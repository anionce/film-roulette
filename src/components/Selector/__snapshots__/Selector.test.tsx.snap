// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correcly 1`] = `
<div>
  <div
    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary root-element css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
  >
    <div
      aria-controls=":r0:"
      aria-expanded="false"
      aria-haspopup="listbox"
      class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
      role="combobox"
      tabindex="0"
    >
      acción
    </div>
    <input
      aria-hidden="true"
      aria-invalid="false"
      class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
      tabindex="-1"
      value="acción"
    />
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
      data-testid="ArrowDropDownIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden="true"
      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        class="css-ihdtdm"
      >
        <span
          class="notranslate"
        >
          ​
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
